### ISSUE TITLE
Export simulate.gam?

### ISSUE THREAD

**Issue by florianhartig** [mfasiolo/mgcViz#76]
Sorry for another one related to https://github.com/florianhartig/DHARMa/issues/309:

Currently, mgcViz:::simulate.gam is not exported, which kind of avoids to overwrite mgcv::simulate.gam, but on the other hand has the disadvantage that when wanting to use this in DHARMa or any other package, I will have to access via ::: which is discouraged (or forbidden?) by CRAN.

What I would suggest is to:

a) either export mgcViz:::simulate.gam, thus overwriting mgcv::simulate.gam, which, however, could also lead to issues

b) export the function under a new name. e.g. mgcViz::simulateGam()

I think option b) has absolutely no costs, and could be implemented on top of the existing code.

**Comment by gavinsimpson** [comment]
Hi @florianhartig There isn't an `mgcv:::simulate.gam`, if `simulate()` works for a model fit by `gam()` or `bam()` it's because those objects inherit from classes `"glm"` and `"lm"` and as such you're getting the base R `simulate`.

I forget the details of how S3 works when importing vs depending, but `mgcViz:::simulate.gam` shouldn't be exported from the package namespace. You may need to depend on {mgcViz} to have the method registered and be called when method dispatch occurs. As such, it would make more sense to get `mgcViz:::simulate.gam` into a more bare bones package with fewer dependencies etc than {mgcViz}.

The major downside of changing the name to `simulateGam` and exporting that is that calling `simulate(my_gam_model)` would no longer work for extended and general families supported by {mgcv}.

**Comment by florianhartig** [comment]
Hi Gavin, if there is no simulate.gam, what is this https://github.com/mfasiolo/mgcViz/blob/9bf589ee5c3e2659718212fa6b70a0b9c3ab9468/R/simulate_gam.R#L43

**Comment by gavinsimpson** [comment]
Note that I wrote that there is no `mgcv:::simulate.gam` as in nothing in **mgcv**, which you claimed, so there's nothing for the version in {mgcViz} to over write, which was partly the point of your issue.  Obviously there is a method in {mgcViz}.

**Comment by mfasiolo** [comment]
Hi Both, 

 @florianhartig is simulate.gam() the only function you are interested in using from `mgcViz`? If so, I agree with Gavin that it would be better to use a barebone package with fewer dependencies than `mgcViz`, such as `mgcvUtils`.

**Comment by gavinsimpson** [comment]
@florianhartig the other option is to soft depend on {mgcViz}; if you're dealing with an object that inherits from class "gam" and the family inherits from class "extended.family", check that {mgcViz} is available and can be loaded via `require()`, if it can't throw an error telling user they need to install {mgcViz}. That way you can just put {mgcViz} in Suggests.

Either way, you don't want to be exporting S3 methods from packages, and duplicating code in {mgcViz} is suboptimal.

**Comment by florianhartig** [comment]
Hi guys,

first of all, @gavinsimpson, sorry, my bad, you are right of course about mgcv::simulate not existing.

Moreover, Gavin, what you suggest in the last message is exactly what I had intended (and already programmed), so what I have implemented at the moment is:

- I check if mgcViz is installed on the system, if so I use it 
- Else I use the standard simulate with a warning

The reason why I wrote the issue, however, is that I was convinced that simulate.gam is not exported by mgcViz, I had somehow tried to access it but couldn't find it, but when looking at your code I now saw that you export, and trying again it now seems to work perfectly.

So, I think this issue not not an issue after all, sorry for the mess. A soft dependency is not such an issue for me, so for the moment I will be happy to add mgcViz to suggest in DHARMa. If you outsource the simulations to mgcvUtils at some point, I'll be happy to switch!
